### PR TITLE
sortperm with dims

### LINKED ISSUE #2061: `sortperm` fails with `dims` keyword
**Describe the bug**

Since Julia 1.9, `Base.sortperm` has accepted a `dims` keyword. This doesn't work with `CuArray`s.

**To reproduce**

The Minimal Working Example (MWE) for this bug:

```julia
julia> using CUDA

julia> xcpu = rand(3,3)
3×3 Matrix{Float64}:
 0.783023   0.240864   0.854814
 0.0488199  0.0840899  0.600029
 0.190407   0.872661   0.386219

julia> sortperm(xcpu; dims=1)
3×3 Matrix{Int64}:
 2  5  9
 3  4  8
 1  6  7

julia> xgpu = cu(xcpu)
3×3 CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}:
 0.783023   0.240864   0.854814
 0.0488199  0.0840899  0.600029
 0.190407   0.872661   0.386219

julia> sortperm(xgpu; dims=1)
ERROR: MethodError: no method matching bitonic_sort!(::Tuple{CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}}; dims::Int64)

Closest candidates are:
  bitonic_sort!(::Any; by, lt, rev) got unsupported keyword argument "dims"
   @ CUDA ~/.julia/packages/CUDA/tVtYo/src/sorting.jl:850

Stacktrace:
 [1] kwerr(::NamedTuple{(:dims,), Tuple{Int64}}, ::Function, ::Tuple{CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}, CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}})
   @ Base ./error.jl:165
 [2] sortperm!(I::CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}, c::CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}; initialized::Bool, kwargs::Base.Pairs{Symbol, Int64, Tuple{Symbol}, NamedTuple{(:dims,), Tuple{Int64}}})
   @ CUDA ~/.julia/packages/CUDA/tVtYo/src/sorting.jl:984
 [3] sortperm(c::CuArray{Float32, 2, CUDA.Mem.DeviceBuffer}; kwargs::Base.Pairs{Symbol, Int64, Tuple{Symbol}, NamedTuple{(:dims,), Tuple{Int64}}})
   @ CUDA ~/.julia/packages/CUDA/tVtYo/src/sorting.jl:989
 [4] top-level scope
   @ REPL[5]:1
 [5] top-level scope
   @ ~/.julia/packages/CUDA/tVtYo/src/initialization.jl:185
```

<details><summary>Manifest.toml</summary>
<p>

```
(@v1.9) pkg> st CUDA
Status `~/.julia/environments/v1.9/Project.toml`
  [052768ef] CUDA v4.4.0

[[deps.CUDA]]
deps = ["AbstractFFTs", "Adapt", "BFloat16s", "CEnum", "CUDA_Driver_jll", "CUDA_Runtime_Discovery", "CUDA_Runtime_jll", "ExprTools", "GPUArrays", "GPUCompiler", "KernelAbstractions", "LLVM", "LazyArtifacts", "Libdl", "LinearAlgebra", "Logging", "Preferences", "Printf", "Random", "Random123", "RandomNumbers", "Reexport", "Requires", "SparseArrays", "SpecialFunctions", "UnsafeAtomicsLLVM"]
git-tree-sha1 = "35160ef0f03b14768abfd68b830f8e3940e8e0dc"
uuid = "052768ef-5323-5732-b1bb-66c8b64840ba"
version = "4.4.0"

[[deps.CUDA_Driver_jll]]
deps = ["Artifacts", "JLLWrappers", "LazyArtifacts", "Libdl", "Pkg"]
git-tree-sha1 = "498f45593f6ddc0adff64a9310bb6710e851781b"
uuid = "4ee394cb-3365-5eb0-8335-949819d2adfc"
version = "0.5.0+1"

[[deps.CUDA_Runtime_Discovery]]
deps = ["Libdl"]
git-tree-sha1 = "bcc4a23cbbd99c8535a5318455dcf0f2546ec536"
uuid = "1af6417a-86b4-443c-805f-a4643ffb695f"
version = "0.2.2"

[[deps.CUDA_Runtime_jll]]
deps = ["Artifacts", "CUDA_Driver_jll", "JLLWrappers", "LazyArtifacts", "Libdl", "TOML"]
git-tree-sha1 = "5248d9c45712e51e27ba9b30eebec65658c6ce29"
uuid = "76a88914-d11a-5bdc-97e0-2f5a05c973a2"
version = "0.6.0+0"

[[deps.CUDNN_jll]]
deps = ["Artifacts", "CUDA_Runtime_jll", "JLLWrappers", "LazyArtifacts", "Libdl", "TOML"]
git-tree-sha1 = "75923dce4275ead3799b238e10178a68c07dbd3b"
uuid = "62b44479-cb7b-5706-934f-f13b2eb2e645"
version = "8.9.4+0"
```

</p>
</details>

**Expected behavior**

The same behavior of `sortperm` when using  `CuArray`s as with `Array`s.

**Version info**

Details on Julia:

```
julia> versioninfo()
Julia Version 1.9.3
Commit bed2cd540a1 (2023-08-24 14:43 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 16 × Intel(R) Core(TM) i7-10875H CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, skylake)
  Threads: 1 on 16 virtual cores
```

Details on CUDA:

```
julia> CUDA.versioninfo()
CUDA runtime 12.1, artifact installation
CUDA driver 12.2
NVIDIA driver 535.98.0

CUDA libraries: 
- CUBLAS: 12.1.3
- CURAND: 10.3.2
- CUFFT: 11.0.2
- CUSOLVER: 11.4.5
- CUSPARSE: 12.1.0
- CUPTI: 18.0.0
- NVML: 12.0.0+535.98

Julia packages: 
- CUDA: 4.4.0
- CUDA_Driver_jll: 0.5.0+1
- CUDA_Runtime_jll: 0.6.0+0

Toolchain:
- Julia: 1.9.3
- LLVM: 14.0.6
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4, 6.5, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5
- Device capability support: sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75, sm_80, sm_86

1 device:
  0: NVIDIA GeForce GTX 1650 Ti (sm_75, 3.596 GiB / 4.000 GiB available)
```

**Additional context**

I believe this bug is due to `bitonic_sort!` being used instead of `quicksort!` within `sortperm!`, and `bitonic_sort!` doesn't support a `dims` keyword (?)

### PR DESCRIPTION
**Pull request by xaellison**
Adds sortperm with dims kwarg (and bitonicsort with dims) to solve #2061

### CODE REVIEW

File: src/sorting.jl [hunk 1]
```diff
@@ -578,10 +578,43 @@ end
...
+function extraneous_block(vals :: AbstractArray, dims):: Bool
+    other_linear_index = ((gridDim().z  ÷ blockDim().z) * (blockIdx().y - 1)) + blockIdx().z
+    return other_linear_index > length(vals) ÷ size(vals)[dims]
+end
+
+function extraneous_block(vals, dims) :: Bool
+    return extraneous_block(vals[1], dims)
+end
+
+# methods are defined for Val{1} because using view has 2x speed penalty for 1D arrays
```
Comment (maleadt) [inline src/sorting.jl:596]: We could probably speed up `view` by returning a `CuDeviceArray` when possible, just like we do with `CuArray`

**Comment by maleadt** [comment]
Nice! So we only use quicksort now for `partialsort!`?

File: src/sorting.jl (hunk 1 above)
Comment (xaellison) [reply src/sorting.jl:596]: Are you suggesting a change here or is this an idea for a separate change in the repo?

**Comment by xaellison** [comment]
> Nice! So we only use quicksort now for `partialsort!`?

That's right. However, some preliminary benchmarking suggests bitonic sort is faster than partialsort for comparable input. Maybe we want to use bitonic for everything? This would also let us add methods for `partialsortperm` (even though they would just be doing a full `sortperm`)

**Comment by maleadt** [comment]
> Maybe we want to use bitonic for everything?

I'd be in favor, also because I haven't implemented the new version of dynamic parallelism yet.

File: src/sorting.jl (hunk 1 above)
Comment (maleadt) [reply src/sorting.jl:596]: A separate change.

File: src/sorting.jl [hunk 2]
```diff
@@ -930,29 +976,35 @@ function Base.sort!(c::AnyCuVector, alg::QuickSortAlg; lt=isless, by=identity, r
...
+    lt=isless, by=identity, rev=false)
+
+    function out(k :: OrdinalRange)
+        return copy(c[k])
     end
 
-    quicksort!(c; lt, by, dims)
-    return c
-end
+    # work around disallowed scalar index
```
Comment (maleadt) [inline src/sorting.jl]: You can do `@allowscalar c[k]`

File: src/sorting.jl [hunk 3]
```diff
@@ -986,6 +1042,11 @@ function Base.sortperm!(ix::AnyCuArray{T}, A::AnyCuArray; initialized=false, kwa
...
 end
 
-function Base.sortperm(c::AnyCuArray; kwargs...)
+function Base.sortperm(c::AnyCuVector; kwargs...)
     sortperm!(CuArray(1:length(c)), c; initialized=true, kwargs...)
 end
+
+function Base.sortperm(c::AnyCuArray; dims, kwargs...)
+    # Base errors for Matrices without dims arg, we should too
+    sortperm!(reshape(CuArray(1:length(c)), size(c)), c; initialized=true, dims=dims, kwargs...)
```
Comment (maleadt) [inline src/sorting.jl]:
```suggestion
    sortperm!(reshape(CuArray(1:length(c)), size(c)), c; initialized=true, dims, kwargs...)
```

**Review by maleadt — APPROVED** [review]
LGTM, thanks!

**Comment by maleadt** [comment]
Some of the kernel launches added here exceed device resources, e.g., during this CI run: https://buildkite.com/julialang/cuda-dot-jl/builds/4942#018ee5fc-7926-4b0a-a855-11039bda2b72

```
base/sorting: Error During Test at /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/test/base/sorting.jl:313
  Test threw exception
  Expression: check_sort!(Int32, (2, 2, 50000); dims = 3, rev = true)
  CUDA error: too many resources requested for launch (code 701, ERROR_LAUNCH_OUT_OF_RESOURCES)
  Stacktrace:
    [1] throw_api_error(res::CUDA.cudaError_enum)
      @ CUDA /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/lib/cudadrv/libcuda.jl:30
    [2] check
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/lib/cudadrv/libcuda.jl:37 [inlined]
    [3] cuLaunchKernel
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/lib/utils/call.jl:30 [inlined]
    [4] #930
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/lib/cudadrv/execution.jl:66 [inlined]
    [5] macro expansion
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/lib/cudadrv/execution.jl:33 [inlined]
    [6] macro expansion
      @ ./none:0 [inlined]
    [7] pack_arguments(::CUDA.var"#930#931"{Bool, Int64, CuStream, CuFunction, CuDim3, CuDim3}, ::CUDA.KernelState, ::CuDeviceArray{Int32, 3, 1}, ::Int32, ::Int32, ::Int32, ::Int32)
      @ CUDA ./none:0
    [8] launch(::CuFunction, ::CUDA.KernelState, ::CuDeviceArray{Int32, 3, 1}, ::Int32, ::Int32, ::Int32, ::Int32; blocks::Tuple{Int64, Int64, Int64}, threads::Tuple{Int64, Int64}, cooperative::Bool, shmem::Int64, stream::CuStream)
      @ CUDA /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/lib/cudadrv/execution.jl:59
    [9] launch
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/lib/cudadrv/execution.jl:52 [inlined]
   [10] #936
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/lib/cudadrv/execution.jl:175 [inlined]
   [11] macro expansion
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/lib/cudadrv/execution.jl:135 [inlined]
   [12] macro expansion
      @ ./none:0 [inlined]
   [13] convert_arguments
      @ ./none:0 [inlined]
   [14] #cudacall#935
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/lib/cudadrv/execution.jl:177 [inlined]
   [15] cudacall
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/lib/cudadrv/execution.jl:173 [inlined]
   [16] macro expansion
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/src/compiler/execution.jl:266 [inlined]
   [17] macro expansion
      @ ./none:0 [inlined]
   [18] call(::CUDA.HostKernel{typeof(CUDA.BitonicSortImpl.comparator_small_kernel), Tuple{CuDeviceArray{Int32, 3, 1}, Int32, Int32, Int32, Int32, typeof(identity), typeof(isless), Val{true}, Val{3}}}, ::CuDeviceArray{Int32, 3, 1}, ::Int32, ::Int32, ::Int32, ::Int32, ::typeof(identity), ::typeof(isless), ::Val{true}, ::Val{3}; call_kwargs::@Kwargs{threads::Tuple{Int64, Int64}, blocks::Tuple{Int64, Int64, Int64}, shmem::Int64})
      @ CUDA ./none:0
   [19] call
      @ ./none:0 [inlined]
   [20] (::CUDA.HostKernel{typeof(CUDA.BitonicSortImpl.comparator_small_kernel), Tuple{CuDeviceArray{Int32, 3, 1}, Int32, Int32, Int32, Int32, typeof(identity), typeof(isless), Val{true}, Val{3}}})(::CuArray{Int32, 3, CUDA.Mem.DeviceBuffer}, ::Int32, ::Int32, ::Int32, ::Int32, ::Function, ::Function, ::Val{true}, ::Val{3}; threads::Tuple{Int64, Int64}, blocks::Tuple{Int64, Int64, Int64}, kwargs::@Kwargs{shmem::Int64})
      @ CUDA /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/src/compiler/execution.jl:389
   [21] bitonic_sort!(c::CuArray{Int32, 3, CUDA.Mem.DeviceBuffer}; by::Function, lt::Function, rev::Bool, dims::Int64)
      @ CUDA.BitonicSortImpl /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/src/sorting.jl:948
   [22] bitonic_sort!
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/src/sorting.jl:902 [inlined]
   [23] #sort!#1224
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/src/sorting.jl:988 [inlined]
   [24] sort!
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/src/sorting.jl:987 [inlined]
   [25] #sort!#1225
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/src/sorting.jl:992 [inlined]
   [26] check_sort!(T::Type, N::Tuple{Int64, Int64, Int64}, f::Function; kwargs::@Kwargs{dims::Int64, rev::Bool})
      @ Main /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/test/base/sorting.jl:198
   [27] check_sort!
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/test/base/sorting.jl:196 [inlined]
   [28] macro expansion
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/test/base/sorting.jl:313 [inlined]
   [29] macro expansion
      @ ~/.cache/julia-buildkite-plugin/julia_installs/bin/linux/x64/1.10/julia-1.10-latest-linux-x86_64/share/julia/stdlib/v1.10/Test/src/Test.jl:669 [inlined]
   [30] macro expansion
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/test/base/sorting.jl:313 [inlined]
   [31] macro expansion
      @ ~/.cache/julia-buildkite-plugin/julia_installs/bin/linux/x64/1.10/julia-1.10-latest-linux-x86_64/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [32] macro expansion
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/test/base/sorting.jl:283 [inlined]
   [33] macro expansion
      @ ~/.cache/julia-buildkite-plugin/julia_installs/bin/linux/x64/1.10/julia-1.10-latest-linux-x86_64/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [34] top-level scope
      @ /var/lib/buildkite-agent/builds/gpuci-16/julialang/cuda-dot-jl/test/base/sorting.jl:281
```

**Pull request merged**